feat(keymap): add `force` to control whether to override present mapping

echasnovski opened the pull request:
Problem: No straightforward way to create a mapping only if there
  doesn't exist one already. This is at least useful for plugins that
  want to be careful to not override user's mappings.

  The `vim.fn.maparg()` does not differentiate between buffer and global
  mappings, making it not a reliable source of truth for global mapping.

  The potential `vim.keymap.get()` might improve things for users (will
  require only something like `if not vim.keymap.get(...) then` instead
  of having to manually iterate), but being able to `force = false`
  still feels like a good improvement.

Solution: Add `opts.force` which if `false` will not create new mapping
  if there already exists one for the same mode+lhs pair.

---

Related: #29464

justinmk commented:
Will this "silently fail"? That seems like it will increase troubleshooting burden. The response to that would be a need for logging or some other indicator, but that could all be avoided by checking `if not vim.keymap.get(...)` or raising an error (`unique`).

echasnovski commented:
> Will this "silently fail"?

What "this"?

`vim.keymap.set('n', lhs, rhs, { force = false })` will not create a mapping if Normal mode mapping for `lhs` already exists. Yes, without extra message, as this is usually how I'd expect "non-forced mappings" to work.

There is no fundamentally more troubleshooting burden than with regular (forced) mappings: forced will override existing mapping (without extra message), non-forced will keep it untouched.

echasnovski committed:
feat(keymap): add `force` to control whether to override present mapping

Problem: No straightforward way to create a mapping only if there
  doesn't exist one already. This is at least useful for plugins that
  want to be careful to not override user's mappings.

  The `vim.fn.maparg()` does not differentiate between buffer and global
  mappings, making it not a reliable source of truth for global mapping.

  The potential `vim.keymap.get()` might improve things for users (will
  require only something like `if not vim.keymap.get(...) then` instead
  of having to manually iterate), but being able to `force = false`
  still feels like a good improvement.

Solution: Add `opts.force` which if `false` will not create new mapping
  if there already exists one for the same mode+lhs pair.

justinmk commented:
> 2\. I don't see why this can't be at least first done as a part of `vim.keymap.set()` to resolve most of the issues (i.e. when a mapping can be or is created via Lua), and then maybe achieve 100% result by porting this flag to `nvim_set_keymap()` and `nvim_buf_set_keymap()`.

because it's confusing, misses out on core features such as `:map` visualization, and is tech debt.
